### PR TITLE
Implement textTransform

### PR DESCRIPTION
**Pull request by asklar**
Implement the `textTransform` prop: none/lowercase/uppercase/capitalize
The prop has some interesting behavior that requires us to differentiate between `none` and `undefined` namely that nested `<Text>` elements will inherit their parent `textTransform` if they don't have one defined themselves.

This is clunkier than I'd like it to be because of the ordering of operations between:
- Creating a view
- Setting its `text` property
- Setting its or its parent `textTransform` property
- Adding the view to its parent's tree

The ordering makes it necessary for us to keep state and calculate things instead of doing the transform in-place when setting the text or transform.

Added an RNTester section in the `<Text>` page which I copied from the iOS/Android examples.

image

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5672)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      case TextTransform::Undefined:
+        [[fallthrough]];
+      case TextTransform::None:
+        return m_originalText;
+      case TextTransform::Uppercase:
+        return boost::to_upper_copy(m_originalText);
+      case TextTransform::Lowercase:
+        return boost::to_lower_copy(m_originalText);
+      case TextTransform::Capitalize: {
+        PWSTR str = new wchar_t[m_originalText.length() + 1];
```
Comment (jonwis) [inline vnext/Microsoft.ReactNative/Utils/TransformableText.h:23]: Consider `auto str = std::make_unique<wchar_t[]>(originalText.length() + 1])` here ... https://en.cppreference.com/w/cpp/string/basic_string/data also suggests that you can just allocate & size the output `std::wstring` carefully, as long `LCMapStringW` correctly null-terminates the buffer exactly at `data()[size()]`.

File: vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h [hunk 2]
```diff
@@ -3,17 +3,52 @@
...
+      std::wstring text(run.Text().c_str());
+      m_transformableText.m_originalText = text;
+      run.Text(m_transformableText.TransformText());
+    } else if (auto span = childSNB.GetView().try_as<xaml::Documents::Span>()) {
+      auto &childVTSN = static_cast<VirtualTextShadowNode &>(child);
+      auto transform = childVTSN.m_transformableText.m_textTransform;
+
+      for (const auto &i : span.Inlines()) {
+        if (auto run = i.try_as<xaml::Documents::Run>()) {
+          std::wstring text(run.Text().c_str());
```
[the remark this answers is not in the document]
Comment (asklar) [reply vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h:29]: yeah, VS 16.7 seems to have a bug where XAML properties aren't showing in the auto/locals window (and calling the Text() method in immediate doesn't work) so I was using this to debug (hence why this is marked draft ;-) )

[the remark this answers is not in the document]
Comment (jonwis) [reply vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h:29]: Ignore the rest of my fb then. :)

[the remark this answers is not in the document]
Comment (asklar) [reply vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h:29]: nah I'm always game to learn something :) I hadn't considered using `make_unique<wchar_t[]>` for ex, thanks for suggesting it :)

File: vnext/src/RNTester/js/examples-win/Text/TextExample.windows.tsx [hunk 3]
```diff
@@ -79,6 +79,39 @@ export class TextExample extends React.Component<{}> {
 
     return (
       <RNTesterPage>
+        <RNTesterBlock title="textTransform">
```
Comment (NickGerleman) [inline vnext/src/RNTester/js/examples-win/Text/TextExample.windows.tsx:82]: Looks like this is from the Android TextExample page.

The examples-win directory was mostly meant for custom windows examples, but TextExample was a bit special since long ago it converted from Flow to Typescript, forked, and stopped being kept up to date. There we're a number of files like this that have been kind of corroded by time before we had tooling to automate three-way-merge of our forked changes.

Consider removing this fork of TextExample and adding back the Flow-based Android one (with any needed patching) to js/examples. That way it will be kept up to date automatically. ACoates and I have been opoortunistically killing the Typescript forks of Flow files when we come across them, and have often found gaps when we do.

Comment (asklar) [reply vnext/src/RNTester/js/examples-win/Text/TextExample.windows.tsx:82]: Filed #5678

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#pragma once
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Utils/TransformableText.h:1]: Nit: missing copyright header

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <boost/algorithm/string.hpp>
+
+namespace react::uwp {
+struct TransformableText final {
+  enum class TextTransform { Undefined, None, Uppercase, Lowercase, Capitalize };
+
+  TextTransform m_textTransform{TextTransform::Undefined};
+  std::wstring m_originalText{};
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Utils/TransformableText.h:10]: Nit: Do we use m_ for public struct members? Used to seeing capital members without prefixes in other parts of the code.

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        auto str = std::make_unique<wchar_t[]>(m_originalText.length() + 1);
+        const int nChars = LCMapStringW(
+            LOCALE_NAME_USER_DEFAULT,
+            LCMAP_TITLECASE,
+            m_originalText.c_str(),
+            static_cast<int>(m_originalText.length()),
+            str.get(),
+            static_cast<int>(m_originalText.length() + 1));
+
+        return std::move(str.get());
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Utils/TransformableText.h:32]: We're doing an extra copy from the c-string buffer to wstring. Would reccomend the alternative solution mentioned in the other comment of creating a wstring, resizing it to the desired text length, then using the non-const ".data()" for C string interop.

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <boost/algorithm/string.hpp>
+
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Utils/TransformableText.h:4]: There's been some conversation about wanting to use less Boost, as we should eventually be rid of Folly in favor of raw JSI and may be able to remove it. to_upper_copy, etc can be pretty easily implemented without dependencies using std::transform and to_upper, etc. Consider that to prevent future cleanup.

File: vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h [hunk 8]
```diff
@@ -3,17 +3,50 @@
...
+
 class VirtualTextViewManager : public ViewManagerBase {
   using Super = ViewManagerBase;
 
  public:
   VirtualTextViewManager(const std::shared_ptr<IReactInstance> &reactInstance);
 
   const char *GetName() const override;
+  facebook::react::ShadowNode *createShadow() const override {
+    return new VirtualTextShadowNode();
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h:48]: I know you aren't the one who added this interface, but this really shouldn't be using a raw pointer 😕

File: vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h [hunk 9]
```diff
@@ -3,17 +3,50 @@
...
+#include <Utils/TransformableText.h>
 #include <Views/FrameworkElementViewManager.h>
 
 namespace react::uwp {
 
+struct VirtualTextShadowNode final : public ShadowNodeBase {
+  using Super = ShadowNodeBase;
+  TransformableText m_transformableText{};
+  void AddView(ShadowNode &child, int64_t index) override {
+    auto &childSNB = static_cast<ShadowNodeBase &>(child);
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Views/VirtualTextViewManager.h:18]: Nit: could you add whitespace before the start of the function? Some of this is a bit hard to parse out as this large of a block.

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h (hunk 6 above)
Comment (NickGerleman) [reply vnext/Microsoft.ReactNative/Utils/TransformableText.h:32]: I think you intended to avoid the copy by doing a move of the raw pointer, but note wstring doesn't do anything special with rvalues to c string pointers https://en.cppreference.com/w/cpp/string/basic_string/basic_string

Doing a move of the raw pointer probably isn't what you want in that case in general though since it won't clear out the unique_ptr. There's a unique_ptr method like "take" or something like that which does though.

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h (hunk 7 above)
Comment (JunielKatarn) [reply vnext/Microsoft.ReactNative/Utils/TransformableText.h:4]: I agree in principle, though, Boost is already a hard dependency for React Native.
Depending on the use-case (don't know enough about this one) there are valid scenarios for leveraging boost, as long as everything stays within the ABI boundary.

Comment (asklar) [reply vnext/Microsoft.ReactNative/Utils/TransformableText.h:4]: > There's been some conversation about wanting to use less Boost, as we should eventually be rid of Folly in favor of raw JSI and may be able to remove it. to_upper_copy, etc can be pretty easily implemented without dependencies using std::transform and to_upper, etc. Consider that to prevent future cleanup.

I considered that but globalization is hard. For example the uppercase of the German letter ß is SS, so you can't really do letter-by-letter. I'll switch everything to LCMapString though since that makes this more uniform and I don't trust that boost might do the right thing for the locale

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h (hunk 5 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/Utils/TransformableText.h:10]: I've seen both (e.g. OInstance uses m_) but I don't feel strongly one way or another

File: vnext/Microsoft.ReactNative/Utils/TransformableText.h (hunk 6 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/Utils/TransformableText.h:32]: > We're doing an extra copy from the c-string buffer to wstring. Would reccomend the alternative solution mentioned in the other comment of creating a wstring, resizing it to the desired text length, then using the non-const ".data()" for C string interop with LCMapStringW.

The problem with using data() is that yes, you can write to it but the size member of the `std::string` doesn't change so effectively the string is empty. The solution is to use `resize` instead of `reserve` to ensure the size is set before/after
